cicerone/spec/reference_resolver.py

```diff
@@ -0,0 +1,200 @@
+
+References:
+- OpenAPI 3.x Reference Resolution: https://spec.openapis.org/oas/v3.1.0#reference-object
+- JSON Pointer: https://datatracker.ietf.org/doc/html/rfc6901
+"""
+
+from __future__ import annotations
+
+import typing
+
+from cicerone.spec import reference as spec_reference
+
+if typing.TYPE_CHECKING:
+    from cicerone.spec import openapi_spec as spec_openapi
+
+
+class ReferenceResolver:
+    """Resolves references in OpenAPI specifications.
+
+    This class provides methods to resolve both internal and external references,
+    navigate the reference graph, and detect circular references.
+    """
+
+    def __init__(self, spec: spec_openapi.OpenAPISpec) -> None:
+        """Initialize the reference resolver.
+
+        Args:
+            spec: The OpenAPI specification to resolve references in
+        """
+        self.spec = spec
+        self._resolution_stack: list[str] = []
+
+    def resolve_reference(
+        self,
+        ref: spec_reference.Reference | str,
+        follow_nested: bool = True,
+    ) -> typing.Any:
+        """Resolve a reference to its target object.
+
+        Args:
+            ref: Reference object or reference string (e.g., '#/components/schemas/User')
+            follow_nested: If True, recursively resolves nested references
+
+        Returns:
+            The target object that the reference points to
+
+        Raises:
+            ValueError: If the reference cannot be resolved
+            RecursionError: If a circular reference is detected
+
+        Example:
+            >>> resolver = ReferenceResolver(spec)
+            >>> user_schema = resolver.resolve_reference('#/components/schemas/User')
+        """
+        # Convert string to Reference object if needed
+        if isinstance(ref, str):
+            ref = spec_reference.Reference(ref=ref)
+
+        # Check for circular references
+        if ref.ref in self._resolution_stack:
+            raise RecursionError(f"Circular reference detected: {' -> '.join(self._resolution_stack + [ref.ref])}")
+
+        # Add to resolution stack for circular reference detection
+        self._resolution_stack.append(ref.ref)
+
+        try:
+            # Currently only support local references
+            if ref.is_external:
+                raise ValueError(f"External references are not yet supported: {ref.ref}")
+
+            # Resolve local reference
+            target = self._resolve_local_reference(ref)
+
+            # If the target is itself a reference and we should follow it
+            if follow_nested and spec_reference.Reference.is_reference(target):
+                nested_ref = spec_reference.Reference.from_dict(target)
+                return self.resolve_reference(nested_ref, follow_nested=True)
+
+            return target
+
+        finally:
+            # Remove from resolution stack
+            self._resolution_stack.pop()
+
+    def _resolve_local_reference(self, ref: spec_reference.Reference) -> typing.Any:
```

Comment:
What we want this to do is return the correct spec object from all the objects under `cicerone/spec/` - without doing this the package itself is useless as we always want to return the rich objects.

I feel like we need some sort of mapping between the objects and the types and then here instead of returning just dictionaries we should always return one of our own spec models.